internal/integration/workloadproxy/workloadproxy.go

```diff
@@ -0,0 +1,668 @@
+								ConfigMap: &corev1.ConfigMapVolumeSource{
+									LocalObjectReference: corev1.LocalObjectReference{
+										Name: identifier,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	if _, err = kubeClient.AppsV1().Deployments(namespace).Create(ctx, deployment, metav1.CreateOptions{}); err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			require.NoErrorf(t, err, "failed to create deployment %q in namespace %q", identifier, namespace)
+		}
+
+		// if the deployment already exists, update it with the new spec
+		deployment, err = kubeClient.AppsV1().Deployments(namespace).Update(ctx, deployment, metav1.UpdateOptions{})
+		if !apierrors.IsNotFound(err) {
+			require.NoError(t, err)
+		}
+	}
+
+	services := make([]*corev1.Service, 0, numServices)
+
+	for i := range numServices {
+		svcIdentifier := fmt.Sprintf("%s-s%02d", identifier, i)
+		port := firstPort + i
+
+		service := corev1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      svcIdentifier,
+				Namespace: namespace,
+				Annotations: map[string]string{
+					constants.ExposedServicePortAnnotationKey:  strconv.Itoa(port),
+					constants.ExposedServiceLabelAnnotationKey: svcIdentifier,
+					constants.ExposedServiceIconAnnotationKey:  icon,
+				},
+			},
+			Spec: corev1.ServiceSpec{
+				Selector: map[string]string{
+					"app": identifier,
+				},
+				Ports: []corev1.ServicePort{
+					{
+						Port:       80,
+						TargetPort: intstr.FromInt32(80),
+					},
+				},
+			},
+		}
+
+		if usePrefix := i%2 == 0; usePrefix {
+			service.Annotations[constants.ExposedServicePrefixAnnotationKey] = strings.ReplaceAll(svcIdentifier, "-", "")
+		}
+
+		services = append(services, &service)
+
+		_, err = kubeClient.CoreV1().Services(namespace).Create(ctx, &service, metav1.CreateOptions{})
+		if !apierrors.IsAlreadyExists(err) {
+			require.NoError(t, err)
+		}
+	}
+
+	// assert that all pods of the deployment are ready
+	require.EventuallyWithT(t, func(collect *assert.CollectT) {
+		dep, depErr := kubeClient.AppsV1().Deployments(namespace).Get(ctx, deployment.Name, metav1.GetOptions{})
+		if errors.Is(depErr, context.Canceled) || errors.Is(depErr, context.DeadlineExceeded) {
+			require.NoError(collect, depErr)
+		}
+
+		if !assert.NoError(collect, depErr) {
+			logger.Error("failed to get deployment", zap.String("namespace", deployment.Namespace), zap.String("name", deployment.Name), zap.Error(depErr))
+
+			return
+		}
+
+		if !assert.Equal(collect, int(dep.Status.ReadyReplicas), numReplicas) {
+			logger.Debug("deployment does not have expected number of replicas",
+				zap.String("namespace", deployment.Namespace),
+				zap.String("name", deployment.Name),
+				zap.Int("expected", numReplicas),
+				zap.Int("found", int(dep.Status.ReadyReplicas)),
+			)
+		}
+	}, 3*time.Minute, 5*time.Second)
+
+	return deployment, services
+}
+
+func doGzip(t *testing.T, input []byte) []byte {
+	t.Helper()
+
+	var buf strings.Builder
```

Comment:
```suggestion
	var buf bytes.Buffer
```

This would simplify and make things more correct